GuiTraceRay from theater ray intersection instead of camera position.

saurtron opened the pull request:
### Work done

* Calculate optimal position to GuiTraceRay from and apply to all GuiTraceRay (the ones with groundOnly=false).

### Explanation

#### Background

* Only applies when using FOV zoom
* Discovered this while testing https://github.com/beyond-all-reason/spring/pull/1774 and https://github.com/beyond-all-reason/spring/pull/1764
* Very similar problem to https://github.com/beyond-all-reason/spring/pull/1754
* The ray origin is too pessimistic, and while this is not a problem when looking down, as the perspective increases it quickly becomes a ray until the other side of the map going back.
* I think this is important since every quad scanned potentially has lots of units in late game, and https://github.com/beyond-all-reason/spring/pull/1764 will multiply scanned quads by (approximately) 3.

#### Implementation

* I define theater as the vertical plane defined by the intersection of camera to far bottom frustum corners with the map maxheight+maxunitaltitude.
*  The theater window is basically the bottom of the frustum intersected with the maximum height + the maximum altitude from airplanes. This defines the closest units we can actually see, even if the camera position is far back.
* Then calculate the theater ray intersection as the intersection of the theater plane with the ray.
* **Note:** For now I don't use the real maxunitaltitude but a hardcoded number, this needs to actually get the maxunitaltitude from somewhere. This comes from unitdefs+some additional logic, and I think can be maintained by the engine, or maybe already does not sure about this. I did check and 500.0 is good for bar atm, but ofc this needs to be improved before merging.
* We could optimize a bit by not doing if camera direction is just looking down with some threshold, but I think the calculations are not too expensive anyways.

#### Remarks

* I think this is the proper logic but of course needs some thought, the idea is tracing from as close as possible to scan less quads.
* Might be I'm reimplementing some functionality already present in engine, please let me know if this is the case. Specially I couldn't find a way to get a plane equation from norm+point.
* Also maybe the "theater" terminology is not the best.

### Before and after

* Top images are from current master, bottom are with the PR applied.
* The white squares are actual quads scanned by GuiTraceRay.
* Note how the top images quickly become a search from the other side of the map going back.
* Note how the corners (represented by squares) I'm drawing go further back than the frustum drawn at minimap, this is because they intersect with ground, but I'm intersecting with max altitude basically.
* Blue square is the camera position, red square the intersection of the ray with the near frustum plane (iirc, I tried this first, but it's also not best).

![rayorigin](https://github.com/user-attachments/assets/07eff63d-82b4-4cf6-b8c1-9985a6549cdd)

### Diagram

![theater](https://github.com/user-attachments/assets/035bf546-b1b9-4ed7-8039-5d73f2d3166c)

saurtron commented:
> Units can be arbitrarily high due to impulse and Lua so there is no such thing as maximum altitude.

Well, in practice atm this is not the case so I think it's worth it to consider. In bar in a typical map with some high areas the max map altitude can be 500 or 1000, units max altitude around 500. Camera can be at 4000 or 5000+.

Even if there was no absolute maximum it could be considered to maintain the highest altitude as an unsynced value, actually I thought the engine might already be maintaining something like this itself but couldn't check properly yet.

saurtron commented:
Added keeping track of unit altitude.

> if it was I still wouldn't want to sacrifice capabilities / behaviour correctness in name of perf, as a matter of principle

Of course I also wouldn't do it, the initial value was just a placeholder while discussing, was just proposing  either we keep track of absolute altitude (as implemented atm) or have some kind of limit set that would be impossible to breach unless the game sets it higher in some way, or maybe the engine had something already, but looking at it doesn't seem to be the case.

Actually if you think about it, maybe the engine does have some higher limit by itself, due to different variables, we just don't know what it is. Of course lua can set it higher, but we can track the limit when lua changes something.

Anyways for now I'm updating the limit pretty much everywhere quadField is getting updated, plus also resetting it on UnitHandler UpdateUnits as that seemed reasonable. I think that should be ok.

The thing i'm not totally sure atm is I'm keeping the limit with `unit->pos.y + unit->radius`, but not totally sure I shouldn't be using `midPos+radius` maybe.

sprunk commented on rts/Sim/Units/UnitHandler.cpp:
```diff
@@ -408,13 +412,15 @@ void CUnitHandler::UpdateUnits()
 {
 	SCOPED_TIMER("Sim::Unit::Update");
 
+	maxUnitAltitude = 0.0f;
```
Does the 0 refer to the water surface or is just a "low" number? Consider picking one of the below accordingly:
```suggestion
	maxUnitAltitude = CGround::GetWaterPlaneLevel();
	maxUnitAltitude = std::numeric_limits<float>::lowest();
```

saurtron replied:
Ok changed this to lowest on Kill(), but GroundMaxHeight otherwise since that's the lowest we should start testing from.

sprunk commented on rts/Game/UI/GuiHandler.cpp:
```diff
@@ -2345,7 +2352,9 @@ Command CGuiHandler::GetCommand(int mouseX, int mouseY, int buttonHint, bool pre
-				const float dist2 = TraceRay::GuiTraceRay(cameraPos, mouseDir, camera->GetFarPlaneDist() * 1.4f, NULL, unit, feature, true);
+				const float viewRange = camera->GetFarPlaneDist() * 1.4f;
+				const float3 rayOrigin = camera->NearTheaterIntersection(mouseDir, viewRange);
+				const float dist2 = TraceRay::GuiTraceRay(rayOrigin, mouseDir, viewRange, NULL, unit, feature, true);
```
Nit: perhaps a good idea to use the modern null while we're touching the line
```suggestion
				const float dist2 = TraceRay::GuiTraceRay(rayOrigin, mouseDir, viewRange, nullptr, unit, feature, true);
```

saurtron replied:
sure

saurtron replied:
done

sprunk reviewed (commented):
Looks decent

sprunk commented:
> The thing i'm not totally sure atm is I'm keeping the limit with unit->pos.y + unit->radius, but not totally sure I shouldn't be using midPos+radius maybe.

It probably should use the top of the selection volume which probably means some nonsense like
```
const auto& selVol = unit->selectionVolume;
const auto top = selVol.GetWorldSpacePos(unit).y + selVol.GetBoundingRadius();
```

Features are also subject to the trace-ray so should be included in the bounds calculation (they also have a selection volume so are handled the ~same as above).

saurtron commented:
> It probably should use the top of the selection volume which probably means some nonsense like
> 
> ```
> const auto& selVol = unit->selectionVolume;
> const auto top = selVol.GetWorldSpacePos(unit).y + selVol.GetBoundingRadius();
> ```

Ok, added this, although I think it'd be good to find a lighter method. Note we don't need exact results but just to be higher than the exact results. Anyways it's not extremely expensive, but we will be looping all units all the time to keep this updated  (otherwise there's no way to reset the altitude and it would just keep going up), plus running it on UnitMove.

> Features are also subject to the trace-ray so should be included in the bounds calculation. Both units and features are derived from the CSolidObject class which is the one that has selectionVolume as a member so the handling would be the same.

Ok added feature support too. I added it inside featureHandler in a similar way to how unitHandler is doing it. I'm doing the altitude reset less often though, as I don't expect this to go up a lot also there is no loop already on features so I had to add one. The good thing is we can check earlier and if the feature max altitude is lower than ground or unit altitude we don't need to loop it.

sprunk commented on rts/Game/Camera.cpp:
```diff
@@ -774,22 +775,30 @@ float3 CCamera::GetMoveVectorFromState(bool fromKeyState) const
 }
 
-float3 CCamera::PointToMaxUnitAltitude(const float3& point, const float rayLength, const float maxAltitude) const
+bool CCamera::TracePointToMaxAltitude(const float3& point, const float rayLength, const float maxAltitude, float3& result) const
```
Perhaps this could return `std::optional <float3>`?

saurtron replied:
done

sprunk commented on rts/Sim/Features/FeatureHandler.cpp:
```diff
@@ -203,6 +220,9 @@ void CFeatureHandler::Update()
 
 		updateFeatures.erase(iter, updateFeatures.end());
 	}
+	if ((gs->frameNum & 62) == 0) {
```
Shouldn't it be 63? With 62 it will do it twice in a row.

saurtron replied:
you're right, changed it

lhog replied:
Where does 62 or 63 come from? Should they be some byproduct of `GAME_SPEED` or `*_SLOWUPDATE_RATE` ?

saurtron replied:
Nowhere really, the number does have to be 2^X-1 I think to actually work properly and not have randomish steps. First I set it to 62 because 2 secs felt right, but didn't really think much about this using a mask instead of a % like other places do.

Not sure what the best practice is for this, didn't want to do the recalculation too much since also I don't foresee features flying a lot, but who knows better be prepared XD.

sprunk commented on rts/Game/Camera.cpp:
```diff
@@ -773,6 +775,47 @@ float3 CCamera::GetMoveVectorFromState(bool fromKeyState) const
+
+float3 CCamera::NearTheaterIntersection(const float3& dir, const float rayLength) const
+{
+	float maxAltitude = std::max(unitHandler.MaxUnitAltitude(), readMap->GetCurrMaxHeight());
+	maxAltitude = std::max(maxAltitude, featureHandler.MaxFeatureAltitude());
```
Both the `unitHandler` max and the `featureHandler` max are initialized with `readMap->GetCurrMaxHeight()`, does that mean it's redundant here? (possibly with `assert`s or comments to document that assumption)
```suggestion
	const float maxAltitude = std::max(unitHandler.MaxUnitAltitude(), featureHandler.MaxFeatureAltitude());
```

saurtron replied:
Yes it is, I'll remove it from here.

saurtron replied:
done

saurtron commented:
> If #1774 is to be merged then everything that updates the respective max should have a visibility check, otherwise the new `/debug` view would leak information about the existence of high-flying stuff within the fog of war.

Yes I was thinking to make that cheat mode only.

sprunk commented:
Also works.

sprunk commented on rts/Sim/Features/FeatureHandler.cpp:
```diff
@@ -203,6 +220,9 @@ void CFeatureHandler::Update()
 
 		updateFeatures.erase(iter, updateFeatures.end());
 	}
+	if ((gs->frameNum & 63) == 0) {
```
> I set it to 62 because 2 secs felt right

How about

```suggestion
	/* Value arbitrary, represents a perf tradeoff:
	 * larger values decrease per-feature overhead over time for the usual state where features don't move vertically,
	 * smaller values decrease traceray overhead if features often enter and leave the stratosphere. */
	static constexpr auto UPDATE_RATE = 2 * GAME_SPEED;
	if (gs->frameNum % UPDATE_RATE == 0) {
```

saurtron replied:
Shouldn't we also change the check above this for destroyed features for consistency too? Otherwise it feels a bit ad-hoc. I did it like this for internal consistency in the class tbh.

Also one concern here is everyone doing checks on frameNum % (n * GAME_SPEED), then all n multipliers of game speed would see perf hit with everyone queueing to do extra work there. So maybe want to add an offset here.

Right now all modules choose "slots" quite randomly resulting in random frames with increased load I guess.

Ideally maybe everyone should query for free/next update slots with some given period stepping so work is distributed among those, probably neither % or & standalone are really good in the end.

Maybe smth like:

```cpp
// example for doing specific slot in 32 (2⁵ frame groups).
static constexpr int PERIOD_STEPPING = 5;
static constexpr int UPDATE_MASK = std::pow(2, PERIOD_STEPPING)-1;
auto slot = engine->getNextSlot(PERIOD_STEPPING);

 if ((frameNum & UPDATE_MASK) % slot) // I think can be done as (frameNum+slot) & UPDATE_RATE
    // do work..
```

I think it could even be improved to sync between different periods. Lua could also use the mechanism, the only issue would be methods who want to do it right now and then with a specific period, but for those probably best would be to actually do it right now somehow, and then conform to the slot assigned by the engine.

Right now here, I think the consistency is important so the function uses the same method for both checks. Later maybe a method like above could be used to properly distribute load among processes.

sprunk replied:
> Shouldn't we also change the check above this for destroyed features for consistency too? Otherwise it feels a bit ad-hoc. I did it like this for internal consistency in the class tbh.

Eventually yes. There are many `x & 31` left all over engine that should mostly become `x % GAME_SPEED` at some point, but some care needs to be taken for each to make sure there isn't a regression. In particular I am unsure if the one above is safe to change this way because running at a slightly slower rate than unit CommandAI slow updates may be on purpose to make sure CAI has an opportunity to process things.

> Right now here, I think the consistency is important so the function uses the same method for both checks.

Sure. I don't mind either way.

> everyone doing checks on frameNum % (n * GAME_SPEED), then all n multipliers of game speed would see perf hit with everyone queueing to do extra work there. So maybe want to add an offset here. Right now all modules choose "slots" quite randomly resulting in random frames with increased load I guess.

Sure. Most of this sort of periodic work is reasonably light though, I'd expect everything just has an offset of 0 but there aren't really spikes every second. Also there may be hidden assumptions so again it would take some care to make sure not to break things (e.g. handler A's periodic update must run right after handler B's and it's not made explicit anywhere, but it sort of happens to work because they just currently have the same offset and period - I remember something like this in resource income stat collection but there may be others).

> Lua could also use the mechanism, the only issue would be methods who want to do it right now and then with a specific period

Lua is a monolith from engine PoV so it is up to Lua to arrange work for individual wupgets, e.g. they could sign up dynamically to the wupgetHandler or use `Script.DelayByFrames`.

saurtron replied:
> care needs to be taken for each to make sure there isn't a regression.

Indeed, any such rework would have to be done with extreme care to understand everything going on.

> Also there may be hidden assumptions so again it would take some care to make sure not to break things (e.g. handler A's periodic update must run right after handler B's and it's not made explicit anywhere, but it sort of happens to work because they just currently have the same offset and period

Current situation is fragile tho, because there might be such a situation happening and might be hard to notice, and any day someone may change some of the periods but then break any such assumptions, like if we decide to change both periods at featureHandler in this PR.

saurtron commented:
diagram of the theater idea:

![theater](https://github.com/user-attachments/assets/035bf546-b1b9-4ed7-8039-5d73f2d3166c)

sprunk commented on rts/Game/Camera.cpp:
```diff
@@ -795,8 +795,16 @@ float3 CCamera::NearTheaterIntersection(const float3& dir, const float rayLength
+	const auto ftl = GetFrustumVert(CCamera::FRUSTUM_POINT_FTL);
+
+	// check the bottom frustum is parallel to ground and lower than the top frustum
+	if ((std::abs(fbl.y-fbr.y) > std::numeric_limits<float>::epsilon()) || (fbl.y >= ftl.y))
```
`epscmp` exists, perhaps use that? Maybe `std::numerlic_limits<T>::epsilon()` could even become the default for its 3rd arg:
https://github.com/beyond-all-reason/spring/blob/a2dd68209e8cb636b3c5a90b07cc929e45f03598/rts/System/SpringMath.h#L141

saurtron commented:
When analyzing the PR with @lhog I noticed the terrible overscanning is only happening when zoom fov is being used (not normal by default, at least in BAR).

In normal situation it's still overscanning when not looking more or less down, although it's not that bad.

Also fixed some error conditions when determininig the closest point. Now I believe it works good.

Still, I think this PR **would benefit from some profiling** before thinking about merge, since under "normal" conditions I believe it's not overscanning too much for most users (just for the ones who use more of a Supreme commander style camera).

Also likely the theaterrayintersection could have a cheaper test to bail out early, but still the main cost in this PR is keeping all altitudes updated.

I posted the updated image at PR description for reference, but posting here as well:

(check the total longitude of the ray, not the width, width is a bit different because angle and position of ray is a bit different in both images, so ray width is affecting differently)

![raystart](https://github.com/user-attachments/assets/bd5bffba-98f1-41b5-8b53-7c3502421551)

saurtron commented:
closing this one until need arises to resurrect or review (basically if fov use becomes used)
